src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,324 @@
-                Contract.ThrowIfNull(asset);
+                while (channelReader.TryRead(out var item))
+                {
+                    await WriteSingleAssetToPipeAsync(
+                        pipeWriter, item.checksum, item.asset, scope, serializer, cancellationToken).ConfigureAwait(false);
+                    foundChecksumCount++;
+                }
+            }
 
-                var kind = asset.GetWellKnownSynchronizationKind();
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // If we weren't canceled, we better have found and written out all the expected assets.
+            Contract.ThrowIfTrue(foundChecksumCount != checksumCount);
+        }
+    }
+
+    private static async ValueTask WriteSingleAssetToPipeAsync(
+        PipeWriter pipeWriter,
+        Checksum checksum,
+        object asset,
+        SolutionAssetStorage.Scope scope,
+        ISerializerService serializer,
+        CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(asset);
+
+        // We're about to send a message.  Write out our sentinel byte to ensure the reading side can detect
+        // problems with our writing.
+        WriteSentinelByte();
+
+        // Write the asset to a temporary buffer so we can calculate its length.  Note: as this is an in-memory
+        // temporary buffer, we don't have to worry about synchronous writes on it blocking on the pipe-writer.
+        // Instead, we'll handle the pipe-writing ourselves afterwards in a completely async fashion.
+
+        using (var _ = GetTempStream(out var tempStream))
+        {
+            WriteAssetToTempStream(tempStream);
+
+            // Write the length of the asset to the pipe writer so the reader knows how much data to read.
+            WriteLength(tempStream.Length);
+
+            // Ensure we flush out the length so the reading side can immediately read the header to determine how much
+            // data to it will need to prebuffer.
+            await pipeWriter.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+            // Now, asynchronously copy the temp buffer over to the writer stream.
+            tempStream.Position = 0;
+            await tempStream.CopyToAsync(pipeWriter, cancellationToken).ConfigureAwait(false);
+        }
+
+        // We flush after each item as that forms a reasonably sized chunk of data to want to then send over the pipe
+        // for the reader on the other side to read.  This allows the item-writing to remain entirely synchronous
+        // without any blocking on async flushing, while also ensuring that we're not buffering the entire stream of
+        // data into the pipe before it gets sent to the other side.
+        await pipeWriter.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+        return;
+
+        void WriteAssetToTempStream(Stream tempStream)
+        {
+            using var writer = new ObjectWriter(tempStream, leaveOpen: true, cancellationToken);
```

Comment:
>new ObjectWriter [](http://example.com/codeflow?start=31&length=16)

It's been almost an hour, so I'll ask here too. Can it be somehow reset to avoid needing to allocate in the loop?

Reply:
yes.  i've been noodling o the idea of a single object-writer and reader.  :)

Reply:
moved to a single object writer and reader.